### PR TITLE
CodeGen RouteTemplate as a RazorCompiledItemAttribute

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Extensions
+{
+    internal class RazorCompiledItemMetadataAttributeIntermediateNode : ExtensionIntermediateNode
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs]: This class should go in the Razor.Langauge assembly

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // [global::...RazorCompiledItemMetadataAttribute(@"{Key}", @"{Value}")]
+            context.CodeWriter.Write("[");
+            context.CodeWriter.Write(AttributeName);
+            context.CodeWriter.Write("(@\"");
+            context.CodeWriter.Write(Key);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs]: What about escaping? I'd suggesting using the extension methods on codewriter that deal with string literals.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                throw new ArgumentNullException(nameof(target));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // [global::...RazorCompiledItemMetadataAttribute(@"{Key}", @"{Value}")]
+            context.CodeWriter.Write("[");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs]: This logic should live in https://github.com/aspnet/Razor/blob/dev/src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributeTargetExtension.cs

File: src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Razor.Hosting
+{
+    /// <summary>
+    /// Defines a key/value metadata pair for the decorated Razor type.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs]: why inherited = false?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs]: Nearly all the attributes in Razor are marked `Inherited = false`. I thought you guys were on to something.

Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs]: The assembly attributes?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs]: Yup:

```
src\Microsoft.AspNetCore.Razor\TagHelpers\HtmlAttributeNameAttribute.cs:11:
[AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
src\Microsoft.AspNetCore.Razor\TagHelpers\HtmlAttributeNotBoundAttribute.cs:11:
[AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
src\Microsoft.AspNetCore.Razor\TagHelpers\HtmlTargetElementAttribute.cs:11:    [AttributeUsage(AttributeTargets.Class,
AllowMultiple = true, Inherited = false)]
src\Microsoft.AspNetCore.Razor\TagHelpers\OutputElementHintAttribute.cs:11:    [AttributeUsage(AttributeTargets.Class,
AllowMultiple = false, Inherited = false)]
src\Microsoft.AspNetCore.Razor\TagHelpers\RestrictChildrenAttribute.cs:15:    [AttributeUsage(AttributeTargets.Class,
Inherited = false, AllowMultiple = false)]
src\Microsoft.AspNetCore.Razor.Language\ProvideRazorExtensionInitializerAttribute.cs:8:
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true, Inherited = false)]
src\Microsoft.AspNetCore.Razor.Runtime\Hosting\RazorCompiledItemAttribute.cs:11:
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true, Inherited = false)]
src\Microsoft.AspNetCore.Razor.Runtime\Hosting\RazorCompiledItemMetadataAttribute.cs:11:
[AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
src\Microsoft.AspNetCore.Razor.Runtime\Hosting\RazorConfigurationNameAttribute.cs:16:
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false, Inherited = false)]
src\Microsoft.AspNetCore.Razor.Runtime\Hosting\RazorExtensionAssemblyNameAttribute.cs:16:
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true, Inherited = false)]
src\Microsoft.AspNetCore.Razor.Runtime\Hosting\RazorLanguageVersionAttribute.cs:16:
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false, Inherited = false)]
src\Microsoft.AspNetCore.Razor.Runtime\Hosting\RazorSourceChecksumAttribute.cs:21:
[AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = true)]
src\Microsoft.CodeAnalysis.Razor.Workspaces\ExportCustomProjectEngineFactoryAttribute.cs:10:
[AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = true)]
test\Microsoft.AspNetCore.Razor.Test.MvcShim\Microsoft.AspNetCore.Mvc.ApplicationParts\ProvideApplicationPartFactoryAtt
ribute.cs:8:    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false)]
test\Microsoft.AspNetCore.Razor.Test.MvcShim\Microsoft.AspNetCore.Mvc.ApplicationParts\RelatedAssemblyAttribute.cs:8:
  [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
test\Microsoft.AspNetCore.Razor.Test.MvcShim\Microsoft.AspNetCore.Mvc.Razor\RazorViewAttribute.cs:8:
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
test\Microsoft.VisualStudio.Editor.Razor.Test.Common\Xunit\ForegroundFactAttribute.cs:10:
[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
test\Microsoft.VisualStudio.Editor.Razor.Test.Common\Xunit\ForegroundTheoryAttribute.cs:10:
[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
```

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
@dotnet-bot build OSX 10.12 Release Build

File: src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs (hunk 4 above)
Comment (NTaylorMullen) [reply src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs]: Lets let this be inherited

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            }
+
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(value));
+            }
+
+            Key = key;
+            Value = value;
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs]: We generally make IR nodes fully mutable. It's a tree so it's handy to mutate them in place.

File: src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs (hunk 4 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemMetadataAttribute.cs]: This attribute is for classes though. That's it should be inherited=true

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs (hunk 5 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Razor.Language/Extensions/RazorCompiledItemMetadataAttributeIntermediateNode.cs]: Okies

**Comment by pranavkm** [comment]
@dotnet-bot build Windows Release x64 Build

**Comment by pranavkm** [comment]
@dotnet-bot test Windows Release x64 Build

**Pull request merged**